cryodrgn/source.py

```diff
@@ -0,0 +1,400 @@
+import os.path
+from collections.abc import Iterable
+from concurrent import futures
+from typing_extensions import reveal_type
+import numpy as np
+import pandas as pd
+from typing import Union, List, Optional
+import logging
+import torch
+
+
+logger = logging.getLogger(__name__)
+
+
+class ImageSourceIterator:
+    def __init__(self, src, chunksize=1000):
+        self.src = src
+        self.chunksize = chunksize
+        self.i = 0
+        self.n = self.src.n
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.i >= self.n:
+            raise StopIteration()
+        indices = np.arange(self.i, min(self.n, self.i + self.chunksize))
+        logger.debug(f"ImageSource returning chunk starting at index {self.i}")
+        images = self.src.images(indices)
+        self.i += self.chunksize
+        return indices, images
+
+
+class ImageSource:
+    @staticmethod
+    def from_file(filepath: str, *args, **kwargs):
+        ext = os.path.splitext(filepath)[-1][1:]
+        assert ext in (
+            "star",
+            "mrcs",
+            "mrc",
+            "txt",
+            "cs",
+        ), f"Unknown file extension {ext}"
+        source = getattr(ImageSource, f"from_{ext}")(filepath, *args, **kwargs)
```

Comment:
The `getattr` trick we can do away with if you find it hacky, though it does serve as good mnemonic so the user only remembers `from_file`.

The args/kwargs were more to assist me so I could mentally shelve arguments I wasn't concerned with while coding this up. The top class needs to have them, but I'd prefer leaving them in there even in the leaf classes. `*args/**kwargs` are going to be all over the place in any extensive code anyway, so maybe there's some setting in `pyright` we can specify to not complain about this?

Reply:
The issue is not that pyright will complain. The issue is that pyright *won't* be able to check the type signatures because it's obscured via the kwargs passthrough.

Reply:
True. And as I was re-reading the code - it does obscure understanding of the code ("Does a `MRCFileSource` ever use `datadir`?", "What classes can leverage multithreading?", for example).

I've gone ahead and removed the args/kwargs stuff from everywhere in `source.py` and am using explicit arguments everywhere now. Removal of `getattr` negates my earlier declaration that "*The top class needs to have them*".

`from_file` is still the preferred way to use this in `cryoDRGN`.

Implemented in df824346d67c327659f268c6cf7dd5db7e320d48